hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSSchedulerNode.java

```diff
@@ -18,35 +18,66 @@
 
 package org.apache.hadoop.yarn.server.resourcemanager.scheduler.fair;
 
+import com.google.common.annotations.VisibleForTesting;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 import org.apache.hadoop.classification.InterfaceAudience.Private;
 import org.apache.hadoop.classification.InterfaceStability.Unstable;
 import org.apache.hadoop.yarn.api.records.ApplicationAttemptId;
+import org.apache.hadoop.yarn.api.records.ContainerId;
+import org.apache.hadoop.yarn.api.records.Resource;
 import org.apache.hadoop.yarn.server.resourcemanager.rmcontainer.RMContainer;
 import org.apache.hadoop.yarn.server.resourcemanager.rmnode.RMNode;
 import org.apache.hadoop.yarn.server.resourcemanager.scheduler.SchedulerApplicationAttempt;
 import org.apache.hadoop.yarn.server.scheduler.SchedulerRequestKey;
 import org.apache.hadoop.yarn.server.resourcemanager.scheduler.SchedulerNode;
+import org.apache.hadoop.yarn.util.resource.Resources;
 
-import java.util.Collection;
-import java.util.Set;
+import java.util.*;
 import java.util.concurrent.ConcurrentSkipListSet;
 
 @Private
 @Unstable
 public class FSSchedulerNode extends SchedulerNode {
 
   private static final Log LOG = LogFactory.getLog(FSSchedulerNode.class);
+  private static final Comparator<RMContainer> comparator =
+      new Comparator<RMContainer>() {
+    @Override
+    public int compare(RMContainer o1, RMContainer o2) {
+      return Long.compare(o1.getContainerId().getContainerId(),
+          o2.getContainerId().getContainerId());
+    }
+  };
 
   private FSAppAttempt reservedAppSchedulable;
-  private final Set<RMContainer> containersForPreemption =
-      new ConcurrentSkipListSet<>();
+  // Stores preemption list until the container is completed
+  @VisibleForTesting
+  final Set<RMContainer> containersForPreemption =
+      new ConcurrentSkipListSet<>(comparator);
+  // Stores preemption list after the container is completed before assigned
+  @VisibleForTesting
+  final Map<FSAppAttempt, Resource>
+      resourcesPreemptedPerApp = new LinkedHashMap<>();
+  Resource totalResourcesPreempted = Resource.newInstance(0, 0);
 
   public FSSchedulerNode(RMNode node, boolean usePortForNodeName) {
     super(node, usePortForNodeName);
   }
 
+  /**
+   * Total amount of reserved resources including reservations and preempted
+   * containers.
+   * @return total resources reserved
+   */
+  Resource getTotalReserved() {
+    Resource totalReserved = getReservedContainer() != null ?
+        Resources.clone(getReservedContainer().getAllocatedResource()) :
+        Resource.newInstance(0, 0);
```

Comment:
Sorry, what is the point cloning none? I think the current one is more straightforward.